Passkey local authentication behavior change

justin-stephenson opened the pull request:
For IPA specifically, we will no longer fallback to local authentication. This enforces the preferred kerberos preauth mechanism to acquire a kerberos ticket during passkey authentication.

justin-stephenson commented:
I included https://github.com/SSSD/sssd/pull/6696/commits/cc5a5c33dc7a6ddec6ec0ee3bb0b2eb2dbf43085 to be sure CI is green, will rebase and drop this commit once it is merged.

justin-stephenson commented:
> I included [cc5a5c3](https://github.com/SSSD/sssd/commit/cc5a5c33dc7a6ddec6ec0ee3bb0b2eb2dbf43085) to be sure CI is green, will rebase and drop this commit once it is merged.

Rebased.

pbrezina reviewed (changes requested):
Can you add some rationale to the commit message since it does not link any ticket?

Also, can you please rebase the code again?

alexey-tikhonov commented:
@justin-stephenson, is this still "no-backport" or does it actually target 2.9.z?

ikerexxe commented:
> @justin-stephenson, is this still "no-backport" or does it actually target 2.9.z?

It does, but it was opened before 2.9 was created, so at that time it didn't make sense to backport it, thus the label. I've updated it.

ikerexxe commented:
Hi @justin-stephenson,

The design page was updated and it seems near its completion. The overall authentication policy has been changed to include two new options that will manage it. I'd recommend you to read it and pay special attention to the [Authentication method policy](https://github.com/SSSD/sssd.io/pull/65/files#diff-1514d401c941fca92a9476ddeadaadeb5b627f30d8f191ddc71d1af1d19b6884R63) and [sssd's configuration options](https://github.com/SSSD/sssd.io/pull/65/files#diff-1514d401c941fca92a9476ddeadaadeb5b627f30d8f191ddc71d1af1d19b6884R260) sections.

You can update this PR accordingly.

justin-stephenson commented:
> Hi @justin-stephenson,
> 
> The design page was updated and it seems near its completion. The overall authentication policy has been changed to include two new options that will manage it. I'd recommend you to read it and pay special attention to the [Authentication method policy](https://github.com/SSSD/sssd.io/pull/65/files#diff-1514d401c941fca92a9476ddeadaadeb5b627f30d8f191ddc71d1af1d19b6884R63) and [sssd's configuration options](https://github.com/SSSD/sssd.io/pull/65/files#diff-1514d401c941fca92a9476ddeadaadeb5b627f30d8f191ddc71d1af1d19b6884R260) sections.
> 
> You can update this PR accordingly.

Hi @ikerexxe @sumit-bose. 

Is it expected all parts should be implemented as part of this PR? It seems quite more involved than I anticipated, I can work on it but I don't really know how long it will take me (just thinking of handling and testing the different permutations of available auth methods + sssd local auth policy + auth order). 


Here is how I parsed the design page, please correct me if have any incorrect assumptions here.

```
Store authentication methods of user against user object (IPA user-auth-types, what about non-IPA case?)

If online, try only authentication methods which return a kerberos ticket
	* IPA case: IDP, Passkey, Smartcard, password
	* AD case: Smartcard, password
	* LDAP Case: N/A

If offline, craft a local_auth_policy list by reading 'local_auth_policy' option:
	if policy="match": try local auth only for previously stored user authentication methods
	
	else if policy="only": try any/all local authentication methods

	else, read "Enable" and "disable" values to generate a list of allowed local auth methods. (What if certain policies are missing and not listed as "enable" or "disable" here ?)

Add a new prompting configuration section [prompting/order] which can be configured to set the authentication prompt order for any PAM service.

```

sumit-bose commented:
> > Hi @justin-stephenson,
> > The design page was updated and it seems near its completion. The overall authentication policy has been changed to include two new options that will manage it. I'd recommend you to read it and pay special attention to the [Authentication method policy](https://github.com/SSSD/sssd.io/pull/65/files#diff-1514d401c941fca92a9476ddeadaadeb5b627f30d8f191ddc71d1af1d19b6884R63) and [sssd's configuration options](https://github.com/SSSD/sssd.io/pull/65/files#diff-1514d401c941fca92a9476ddeadaadeb5b627f30d8f191ddc71d1af1d19b6884R260) sections.
> > You can update this PR accordingly.
> 
> Hi @ikerexxe @sumit-bose.
> 
> Is it expected all parts should be implemented as part of this PR? It seems quite more involved than I anticipated, I can work on it but I don't really know how long it will take me (just thinking of handling and testing the different permutations of available auth methods + sssd local auth policy + auth order).
> 
> Here is how I parsed the design page, please correct me if have any incorrect assumptions here.
> 
> ```
> Store authentication methods of user against user object (IPA user-auth-types, what about non-IPA case?)
> 
> If online, try only authentication methods which return a kerberos ticket
> 	* IPA case: IDP, Passkey, Smartcard, password
> 	* AD case: Smartcard, password
> 	* LDAP Case: N/A
> 
> If offline, craft a local_auth_policy list by reading 'local_auth_policy' option:
> 	if policy="match": try local auth only for previously stored user authentication methods
> 	
> 	else if policy="only": try any/all local authentication methods
> 
> 	else, read "Enable" and "disable" values to generate a list of allowed local auth methods. (What if certain policies are missing and not listed as "enable" or "disable" here ?)

Hi,

my understanding/idea was a bit different. First, my understanding is that prompting should be the same when online or offline since it might be otherwise irritating for users and might give attackers information about the online/offline state of the system. The `local_auth_policy` controls if additional 'local' methods should be presented to the user other than the ones the backend itself can handle. By default the `krb5` auth provider and so `ipa` and `ad` as well, can only handle Kerberos so 'try only authentication methods which return a kerberos ticket' applies here and this means that by default, i.e. `local_auth_provider = match`, all methods the given KDC supports can be presented to the user both when online or offine.

For the `ldap` auth provider we currently only support password based authentication, so `local_auth_policy = match` would mean to only allow password authentication when being online or offline. 

The other, non-default, values of `local_auth_policy` will add "local" authentication methods which are not supported by the backend and the KDC, e.g. passkey for AD  or LDAP users and this as well both for online and offline. Only in the online case with a Kerberos based auth provider there should be a message for the user, if a "local" method is selected that there will be no Kerberos ticket.

> 
> Add a new prompting configuration section [prompting/order] which can be configured to set the authentication prompt order for any PAM service.
> ```

I think the prompting order should not be part of this PR.

HTH

bye,
Sumit

justin-stephenson commented:
> Hi,
> 
> my understanding/idea was a bit different. First, my understanding is that prompting should be the same when online or offline since it might be otherwise irritating for users and might give attackers information about the online/offline state of the system. The `local_auth_policy` controls if additional 'local' methods should be presented to the user other than the ones the backend itself can handle. By default the `krb5` auth provider and so `ipa` and `ad` as well, can only handle Kerberos so 'try only authentication methods which return a kerberos ticket' applies here and this means that by default, i.e. `local_auth_provider = match`, all methods the given KDC supports can be presented to the user both when online or offine.
> 
> For the `ldap` auth provider we currently only support password based authentication, so `local_auth_policy = match` would mean to only allow password authentication when being online or offline.
> 
> The other, non-default, values of `local_auth_policy` will add "local" authentication methods which are not supported by the backend and the KDC, e.g. passkey for AD or LDAP users and this as well both for online and offline. Only in the online case with a Kerberos based auth provider there should be a message for the user, if a "local" method is selected that there will be no Kerberos ticket.
> 
> > Add a new prompting configuration section [prompting/order] which can be configured to set the authentication prompt order for any PAM service.
> 
> I think the prompting order should not be part of this PR.

Thank you very much for clarification, that answers most of my questions. Does smartcard authentication in IPA/AD grant a kerberos ticket? For testing purpose is shutting down the backend server a sufficient test for smartcard authentication to switch to local smartcard auth ? 
> 
> HTH
> 
> bye, Sumit

sumit-bose commented:
> 
> Thank you very much for clarification, that answers most of my questions. Does smartcard authentication in IPA/AD grant a kerberos ticket? For testing purpose is shutting down the backend server a sufficient test for smartcard authentication to switch to local smartcard auth ?

Hi,

I think you cannot hardcode Smartcard authentication to the IPA/AD provider but you have to look at which types are returned by `krb5_child` in the `pam_data` struct similar to what is done in `pam_eval_prompting_config()`. This will show you which methods are available for the given user to get a Kerberos ticket.

For testing you can remove `/usr/lib64/krb5/plugins/preauth/pkinit.so` or the whole `krb5-pkinit` package from either the client or the server. Then all other methods, except Smartcard, should still be available.

> 
> Also, IIUC only 'local' methods are passkey, smartcard, or password. 2FA and IDP are online only.
> 

I would say the `password` is online only as well, because both `passkey` and `Smartcard` would work on a system with an empty cache if SSSD can read the required data from LDAP. For `password` we need a successful online authentication first to store the password hash for offline authentication.

bye,
Sumit

justin-stephenson commented:
> For the `ldap` auth provider we currently only support password based authentication, so `local_auth_policy = match` would mean to only allow password authentication when being online or offline.
> 

In this case is there any code change needed here when provider is LDAP and `local_auth_policy = match`? Or is this behavior already the default behavior of the ldap provider?

justin-stephenson commented:
> > Thank you very much for clarification, that answers most of my questions. Does smartcard authentication in IPA/AD grant a kerberos ticket? For testing purpose is shutting down the backend server a sufficient test for smartcard authentication to switch to local smartcard auth ?
> 
> Hi,
> 
> I think you cannot hardcode Smartcard authentication to the IPA/AD provider but you have to look at which types are returned by `krb5_child` in the `pam_data` struct similar to what is done in `pam_eval_prompting_config()`. This will show you which methods are available for the given user to get a Kerberos ticket.
> 
I created a function to check the auth types similar to `pam_eval_prompting_config`, If I create an IPA user, and set `--user-auth-type=pkinit` then "smartcard" is not returned in the pam data response list type. Is it expected? (I didn't actually plug a smartcard into my system or do any futher smartcard setup).

sumit-bose commented:
> > > Thank you very much for clarification, that answers most of my questions. Does smartcard authentication in IPA/AD grant a kerberos ticket? For testing purpose is shutting down the backend server a sufficient test for smartcard authentication to switch to local smartcard auth ?
> > 
> > 
> > Hi,
> > I think you cannot hardcode Smartcard authentication to the IPA/AD provider but you have to look at which types are returned by `krb5_child` in the `pam_data` struct similar to what is done in `pam_eval_prompting_config()`. This will show you which methods are available for the given user to get a Kerberos ticket.
> 
> I created a function to check the auth types similar to `pam_eval_prompting_config`, If I create an IPA user, and set `--user-auth-type=pkinit` then "smartcard" is not returned in the pam data response list type. Is it expected? (I didn't actually plug a smartcard into my system or do any futher smartcard setup).

Hi,

yes, this is expected, currently SSSD will only do Smartcard authentication if a Smartcard is available.

bye,
Sumit

justin-stephenson commented:
@sumit-bose This PR is ready for review again, I tested these changes with passkey authentication however I'm still having issues getting smartcard auth working on my local system so I may ask if you wouldn't mind performing the smartcard testing.

We will want tests for different `local_auth_policy` configurations but I suppose they need to wait for, or added to https://github.com/SSSD/sssd/pull/6634 ?

justin-stephenson commented:
> 
> There wasn't any message to connect the passkey or to input the PIN, and I was able to authenticate because the FIDO2 key LED was blinking and I decided to touch it.

Hi, thank you for testing. I do not see any LED blink when I am following the same steps. (btw local_auth_policy = match is default so adding it should not make any difference).

https://paste.centos.org/view/40c63201

justin-stephenson commented on src/responder/pam/pamsrv_cmd.c:
```diff
@@ -746,6 +749,150 @@ errno_t filter_responses(struct pam_ctx *pctx,
+                            CONFDB_DOMAIN_LOCAL_AUTH_POLICY,
+                            "match", &local_policy);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Failed to get the confdb local_auth_policys\n");
```
Fixed, it was just a typo there should be no `s`

justin-stephenson commented on src/responder/pam/pamsrv_cmd.c:
```diff
@@ -1136,6 +1285,27 @@ void pam_reply(struct pam_auth_req *preq)
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Local auth policy allowed: smartcard [%s], passkey [%s]\n",
+                             local_sc_auth_allow ? "True" : "False",
+                             local_passkey_auth_allow ? "True" : "False");
```
What should it be?

ikerexxe replied:
I'm reviewing the [Coding Style](https://sssd.io/contrib/coding-style.html) and the definition is broader than I thought. The code is fine so I withdraw this comment

justin-stephenson commented on src/responder/pam/pamsrv_cmd.c:
```diff
@@ -2813,6 +3028,9 @@ static void pam_dom_forwarder(struct pam_auth_req *preq)
 
+    if (tmp_ctx != NULL) {
+        talloc_free(tmp_ctx);
+    }
```
Fixed.

ikerexxe commented:
> Hi, thank you for testing. I do not see any LED blink when I am following the same steps.

I'll check again tomorrow in a clean environment. In any case, IIUIC if the online case prioritizes passkey authentication, then the offline case when `local_auth_policy = match` is set should prioritize the same. I don't see that behaviour in teh steps that you provided.

> (btw local_auth_policy = match is default so adding it should not make any difference).

I tried to note all the steps just in case, but I agree that there shouldn't be any difference.

justin-stephenson commented:
> > Hi, thank you for testing. I do not see any LED blink when I am following the same steps.
> 
> I'll check again tomorrow in a clean environment. In any case, IIUIC if the online case prioritizes passkey authentication, then the offline case when `local_auth_policy = match` is set should prioritize the same. I don't see that behaviour in teh steps that you provided.
> 

You are correct, my mistake. I'm working on a fix.

ikerexxe commented:
Ok, two minor changes to this workflow

>     * Create user and add passkey
> 
>     * Log in and log out (to cache the user)
> 
>     * Make the server unavailable by changing the `ipa_domain` at sssd.conf
> 
>     * Add `local_auth_policy = match` to sssd.conf

As you mentioned, this isn't actually needed as it's already the default.

> 
>     * Restart sssd service
> 
>     * Log in

When I'm requested the PIN I pressed enter (empty string). That's when it fallbacks to passkey authentication, but without any instructions on the screen. I know it has fallback because the FIDO2 key LED is blinking.

ikerexxe commented:
I've tested all authentication mechanisms except for smartcards (I don't have any) and there's a workflow that isn't working as expected. If I log in and input an empty PIN for the passkey, and an empty password, then the LED starts blinking without any prompt. Example:
```
[ci@client ~]$ su - alice@ipa.test
Insert your passkey device, then press ENTER.
Enter PIN:               <- Empty string
Password:                <- Empty string
                         <- At this point the LED starts blinking
```

IIRC, the fallback to passkeys is intended, but the user should get some notification that we'll try this method. This is similar to the concern I raised on Monday.

justin-stephenson commented:
> I've tested all authentication mechanisms except for smartcards (I don't have any) and there's a workflow that isn't working as expected. If I log in and input an empty PIN for the passkey, and an empty password, then the LED starts blinking without any prompt. Example:
> 
> ```
> [ci@client ~]$ su - alice@ipa.test
> Insert your passkey device, then press ENTER.
> Enter PIN:               <- Empty string
> Password:                <- Empty string
>                          <- At this point the LED starts blinking
> ```
> 
> IIRC, the fallback to passkeys is intended, but the user should get some notification that we'll try this method. This is similar to the concern I raised on Monday.

Fallback from empty string at "Enter PIN" to password is expected, but empty string as the password prompt should not fallback to anything IIUC. I will investigate this.

justin-stephenson commented:
> > I've tested all authentication mechanisms except for smartcards (I don't have any) and there's a workflow that isn't working as expected. If I log in and input an empty PIN for the passkey, and an empty password, then the LED starts blinking without any prompt. Example:
> > ```
> > [ci@client ~]$ su - alice@ipa.test
> > Insert your passkey device, then press ENTER.
> > Enter PIN:               <- Empty string
> > Password:                <- Empty string
> >                          <- At this point the LED starts blinking
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > IIRC, the fallback to passkeys is intended, but the user should get some notification that we'll try this method. This is similar to the concern I raised on Monday.
> 
> Fallback from empty string at "Enter PIN" to password is expected, but empty string as the password prompt should not fallback to anything IIUC. I will investigate this.

@sumit-bose 

When fallback to `prompt_password` from `prompt_passkey` occurs in  `src/sss_client/pam_sss.c` in the below code

```
            } else if (pi->passkey_prompt_pin) {
                ret = prompt_passkey(pamh, pi,
                                     _("Insert your passkey device, then press ENTER."),
                                     "");
                /* Fallback to password auth if no PIN was entered */
                if (ret == EIO) {
                    ret = prompt_password(pamh, pi, _("Password: "));
                }
```

This causes the backend to trigger the passkey kerberos preauth flow again, and we get prompted for a passkey pin. 

~~~
/var/log/sssd/krb5_child.log:182:(2023-06-14 16:45:09): [krb5_child[1261605]] [sss_krb5_responder] (0x4000): [RID#10] Got question [passkey].
~~~

Could you help me determine where I can intercept this request and return, or how to prevent this from happening? In the Preauth phase the authtok type is always SSS_AUTHTOK_TYPE_EMPTY so i'm not sure what condition can be used to distinguish this, or can I use pam_set_item in pam_sss.c somehow?

justin-stephenson commented:
@ikerexxe @sumit-bose  aside from my [previous comment](https://github.com/SSSD/sssd/pull/6696#issuecomment-1591974758) which I need some advice/help on how to resolve, this is ready for more thorough review. Especially smartcard authentication flows.

ikerexxe commented:
> @ikerexxe @sumit-bose aside from my [previous comment](https://github.com/SSSD/sssd/pull/6696#issuecomment-1591974758) which I need some advice/help on how to resolve, this is ready for more thorough review. Especially smartcard authentication flows.

The only thing that is missing from my side is the fix for the fallback. Once that is done, and taking into account that I'm unable to test the smartcard authentication, I'll approve the changes.

sumit-bose commented:
Hi,

I think the most straight forward way would be to already out in pam_sss.c if the fallback password prompt returns an empty result as well. In general we do not want empty passwords anyway but so far we said that it is up to the backend and the connected server to decide. However, in this fallback chain where an available secure method is skipped I think we can safely assume that the user is not configured to be allowed to log in with an empty password.

bye,
Sumit

justin-stephenson commented:
> Hi,
> 
> I think the most straight forward way would be to already out in pam_sss.c if the fallback password prompt returns an empty result as well. In general we do not want empty passwords anyway but so far we said that it is up to the backend and the connected server to decide. However, in this fallback chain where an available secure method is skipped I think we can safely assume that the user is not configured to be allowed to log in with an empty password.
> 
> bye, Sumit

Hi,

I am returning PAM_AUTHTOK_ERR  in `get_authtok_for_authentication` when the empty password is detected but it still fallsback to the prompt_passkey prompt again.

            } else if (pi->passkey_prompt_pin) {
                ret = prompt_passkey(pamh, pi,
                                     _("Insert your passkey device, then press ENTER."),
                                     "");
                /* Fallback to password auth if no PIN was entered */
                if (ret == EIO) {
                    ret = prompt_password(pamh, pi, _("Password: "));
                    if (pi->pam_authtok_size == 0) {
                        D(("Empty password failure"));
                        pi->passkey_prompt_pin = false;
                        return PAM_AUTHTOK_ERR;
                    }
                } 

~~~
[justin@fedora cockpit-session-recording]$ ssh testuser1@ipa.test@localhost
(testuser1@ipa.test@localhost) Insert your passkey device, then press ENTER.
(testuser1@ipa.test@localhost) Enter PIN:
(testuser1@ipa.test@localhost) Password: 
(testuser1@ipa.test@localhost) Insert your passkey device, then press ENTER.
~~~

sumit-bose commented:
Hi,

are you sure this is not ssh client which retries? Do you see the same with

    ssh -o NumberOfPasswordPrompts=1  testuser1@ipa.test@localhost

bye,
Sumit

justin-stephenson commented:
> Hi,
> 
> are you sure this is not ssh client which retries? Do you see the same with
> 
> ```
> ssh -o NumberOfPasswordPrompts=1  testuser1@ipa.test@localhost
> ```
> 
> bye, Sumit

Thanks, you are correct it was SSH retrying.

Pushed new changes to this PR, it is ready for review again.

ikerexxe reviewed (approved):
LGTM. Thanks for your hard work Justin! :smile:

sumit-bose commented:
Hi,

thank you for the patches. In my mind I saw the functionality implemented in the common backend code, but doing it in the PAM responder makes even more sense. Currently you are evaluating the `local_auth_policy` option twice, first in `pam_dom_forwarder()` and second in `pam_eval_local_auth_policy()` which is called from `pam_reply()`. I wonder if it would be easier to read it once and store the result in `struct pam_auth_req()`?

I had to apply some changes to make Smartcard authentication work as expected, see `fixes` patch in https://github.com/sumit-bose/sssd/commits/my_pr6696. The changes are mainly about the "CERT_INFO" item of the PAM `resp_list`. This items contains information about the found, matched and mapped certificate for the given user. This is on one hand needed my `pam_sss.so` to decide if a prompt for the Smartcard PIN should be shown (or even the certificate selection first). On the other hand it is needed by `krb5_child` to add the required pre-auth options so that the pkinit plugin can check if pkinit is possible or not. If pkinit is possible `krb5_child` will set `SSS_CERT_AUTH_PROMPTING`, if the backend or KDC cannot handle pkinit `SSS_CERT_AUTH_PROMPTING` is not set.

For the `local_auth_policy = only` case you have to make sure that the "CERT_INFO" is set, if a certificate was found, so the `pam_sss.so` cna ask for the PIN, that why I moved the evaluation of `only` a but further down in `pam_dom_forwarder()`.

In the `match` case, if the backend cannot handle pkinit, i.e. if  `SSS_CERT_AUTH_PROMPTING` is not set, the "CERT_INFO" may not be send to `pam_sss.so` because otherwise the user would be asked for a PIN. This is handled by the changes to `pam_get_auth_types`.

Feel free to pick what you need from the `fixes` patches and modify it as needed because I'm not sure if I always used the best flow to avoid dubplications.

I tested with my WIP patch for the proxy provider as well and it worked well.

bye,
Sumit

sumit-bose commented:
Hi,

I came across another issues, if passkey is enabled but Smartcard authentication is not but possible, the cert info data has to be remove as well. See https://github.com/sumit-bose/sssd/commit/28ae1dfb0560a1294b964d6cb453683c1978ffbc.

I think in the long run it might be even better to set `do_not_send_to_client` to `true` for the cert info data by default and only set it to `false`, i.e. send it to the client, if Smartcard authentication is allowed. But imo for the time being we can continue this way and refactor the code when we have a suitable test coverage for the `local_auth_policy` option.

bye,
Sumit

justin-stephenson commented:
Thank you @sumit-bose I incorporated your fixes as is, including the `do_not_send_cert_info` change. I just added a couple comments in the code to clarify things as you've mentioned here in github.

I also added a fix in `pam_reply()`for  the LDAP provider default (local auth policy = match) to perform password auth only, as stated earlier in this PR comments. Setting the the policy to 'only' or 'enable:passkey' will explicitly enables passkey auth.

> Currently you are evaluating the local_auth_policy option twice, first in pam_dom_forwarder() and second in pam_eval_local_auth_policy() which is called from pam_reply(). I wonder if it would be easier to read it once and store the result in struct pam_auth_req()?

I simplified the code a bit by having both pam functions call pam_eval_local_auth_policy() and I changed this function to return both the auth policy confdb string, and also the `sc_allow` and `passkey_allow` booleans

> @justin-stephenson are you ready for another round of reviews? Or do you want to do anything else with this code?

@ikerexxe I'm waiting for CI results then will add the label back waiting for review.

justin-stephenson commented:
Covscan shows the following warning, but I don't see it related to my patches. Please correct me if i'm mistaken.

~~~
 Warning: Error: CLANG_WARNING:
sssd-pr6696/src/sss_client/pam_sss.c:2966:21: warning[unix.Malloc]: Potential leak of memory pointed to by 'pi.pam_authtok'
# 2964|                   ret = get_authtok_for_authentication(pamh, &pi, flags);
# 2965|                   if (ret != PAM_SUCCESS) {
# 2966|->                     D(("failed to get authentication token: %s",
# 2967|                          pam_strerror(pamh, ret)));
# 2968|                       return ret;
~~~

sumit-bose commented:
Hi,

here are some step which should help to test Smartcard authentication with sssd-ci-containers and no additional hardware or other devices. The idea is to use the same mechanism as the unit and integration tests based on the SoftHSM2 PKCS#11 and use the certificates generated for the tests. I used the `ci-client-devel` image so that all dependencies to call `make test_CA` are available and I guess SoftHSM2 isn't available in the default client image as well.

So start with:
```
git clone https://github.com/SSSD/sssd 
cd sssd
. ./contrib/fedora/bashrc_sssd
reconfig
make test_CA
```

Then you can check your "Smartcard" and export the certificate:
```
# SOFTHSM2_CONF=/root/sssd/x86_64/src/tests/test_CA/softhsm2_one.conf p11tool --list-all
warning: no token URL was provided for this operation; the available tokens are:

pkcs11:model=p11-kit-trust;manufacturer=PKCS%2311%20Kit;serial=1;token=System%20Trust
pkcs11:model=p11-kit-trust;manufacturer=PKCS%2311%20Kit;serial=1;token=Default%20Trust
pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=6855051c589d1676;token=SSSD%20Test%20Token
# SOFTHSM2_CONF=/root/sssd/x86_64/src/tests/test_CA/softhsm2_one.conf p11tool --list-all 'pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=6855051c589d1676;token=SSSD%20Test%20Token'
Object 0:
	URL: pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=6855051c589d1676;token=SSSD%20Test%20Token;id=%C5%54%C9%F8%2C%2A%9D%58%B7%09%21%C1%43%30%41%53%A8%A4%2F%17;object=SSSD%20test%20cert%200001;type=cert
	Type: X.509 Certificate (RSA-2048)
	Expires: Sun Feb  4 11:19:25 2024
	Label: SSSD test cert 0001
	ID: c5:54:c9:f8:2c:2a:9d:58:b7:09:21:c1:43:30:41:53:a8:a4:2f:17
```
The first call will return the URI of the SoftHSM2 device which can be used as an argument to the second call to return its content which includes the URI of the certificate. I use single quotes (') around the URIs to avoid issues with the shell. The certificate URI can be used to export the certificate:
```
# SOFTHSM2_CONF=/root/sssd/x86_64/src/tests/test_CA/softhsm2_one.conf p11tool --export --outder 'pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=6855051c589d1676;token=SSSD%20Test%20Token;id=%C5%54%C9%F8%2C%2A%9D%58%B7%09%21%C1%43%30%41%53%A8%A4%2F%17;object=SSSD%20test%20cert%200001;type=cert' | base64 -w0
MIIE/TCCAuWgAwIBAgIBATANBgkqhkiG9w0BAQsFADA6MQ0wCwYDVQQKDARTU1NEMRIwEAYDVQQLDAlTU1NEIHRlc3QxFTATBgNVBAMMDFNTU0QgdGVzdCBDQTAeFw0yMzA3MTkxMTE5MjVaFw0yNDAyMDQxMTE5MjVaMEExDTALBgNVBAoMBFNTU0QxEjAQBgNVBAsMCVNTU0QgdGVzdDEcMBoGA1UEAwwTU1NTRCB0ZXN0IGNlcnQgMDAwMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANfzGCUs/4Pnh0ZX+Wd2ZDthGDUYO5pmKXOH+N8C9remxJgOefV/3CbKA3HQdSOWkbvakE5pi2F2nUNkZrJOWlJRj/fLv6m+iQWVjS/1eGDfKFTJEr3m5L8u5jgvy5qhnExc9baI/eTO/a8FBCUy4EKMv6OL+1fZmXwIdSV1jtA+1ItlySVeZG+7mMc8rlb96HwTrpQ15v9256pkDMsFlvmDwGW/10yN/95KUz9H/9q7jRkEFrcJGl11p/+A5r2w0ZV3pGbPB6NnjuSE/CTHq3bdlDttmSTQ0jcfz7+wqQzppZlrLCYu+fIKUiCOBmHnMomC/RkUBpD+dJNrqR7c560CAwEAAaOCAQUwggEBMB8GA1UdIwQYMBaAFBg3LrFmVu3491fKzSAkNuCZFletMAkGA1UdEwQCMAAwEQYJYIZIAYb4QgEBBAQDAgWgMCQGCWCGSAGG+EIBDQQXFhVTU1NEIHRlc3QgQ2VydGlmaWNhdGUwHQYDVR0OBBYEFLqZmpTbOgVz8KSsrJXxwzW9+HGZMA4GA1UdDwEB/wQEAwIF4DAdBgNVHSUEFjAUBggrBgEFBQcDAgYIKwYBBQUHAwQwTAYDVR0RBEUwQ4Ehc3NzZC1kZXZlbEBsaXN0cy5mZWRvcmFob3N0ZWQub3Jnhh5odHRwczovL2dpdGh1Yi5jb20vU1NTRC9zc3NkLy8wDQYJKoZIhvcNAQELBQADggIBALNKTUZSwDLS13AVd2oJwkcnAPItwFlJwNBZurnOlkqpddFrGrY1s3AR+1V8ioZc/gy50ccf6oAsfg0M4uv3437BIwHAYS1nif+3fLWtmF6rnleM38G2xU1CTu4tlt5CQ74Lyg9wkL9qeSCxnzQyKdC3O3AkHnU9Go0uVEx6lAsFunFaRNvgcfZb98DVxCONkTvBDBiu4BngySw450kVz8SVboBNgxGyBH89lFH/D37+ATMLXo3lXZJZ0AW3b5jpffpeBlls82VAuK+7Wh43p/+1s0Mh/zEiX8UjRO1FVU0U3IRb/V6HnOTYpty8KUKciPBKUvrfyaHAyYLdbyhmWRl3wJn8NxkeLid7XLulapGxe8GBJ3WvWdaN/xtHICer68Ib8QnqbHXQE2CPsFzUZ846ZpYIlPpOQ7LJQmImAJzTaoGTjlVDWOcg3sVEjDBSw2H49Dac8RBEQI5XbSXQPutfEpB3U7uCKLfBZGf6+wZPws3XiMTtxx9v9apypVJtINBl3SILTWLZnwjhibMRr2MHMGMJOm6zju/pO13VSmZTuHpw/WIbIQrFOybnB8LV4DNlR8NrGaZ8hrjTlY+2wS2NpY2rlz+J6FmU9kdH9CpMqYUBWDrbj4K7nPA9Wgt9/4AL9rKKONtM1ugKxtB/tsO6lX/DFkRaso0fVX3S+Pey
```
The DER output together with `base64 -w0` makes the output in the format the IPA utility expects as argument for the certificate.

Now switch to the IPA server, create a user and add the certificate:
```
# ipa user-add scuser
First name: Sc
Last name: User
-------------------
Added user "scuser"
-------------------
  User login: scuser
  First name: Sc
  Last name: User
  Full name: Sc User
  Display name: Sc User
  Initials: SU
  Home directory: /home/scuser
  GECOS: Sc User
  Login shell: /bin/sh
  Principal name: scuser@IPA.TEST
  Principal alias: scuser@IPA.TEST
  Email address: scuser@ipa.test
  UID: 858600005
  GID: 858600005
  Password: False
  Member of groups: ipausers
  Kerberos keys available: False

# ipa user-add-cert scuser --certificate=MIIE/TCCAuWgAwIBAgIBATANBgkqhkiG9w0BAQsFADA6MQ0wCwYDVQQKDARTU1NEMRIwEAYDVQQLDAlTU1NEIHRlc3QxFTATBgNVBAMMDFNTU0QgdGVzdCBDQTAeFw0yMzA3MTkxMTE5MjVaFw0yNDAyMDQxMTE5MjVaMEExDTALBgNVBAoMBFNTU0QxEjAQBgNVBAsMCVNTU0QgdGVzdDEcMBoGA1UEAwwTU1NTRCB0ZXN0IGNlcnQgMDAwMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANfzGCUs/4Pnh0ZX+Wd2ZDthGDUYO5pmKXOH+N8C9remxJgOefV/3CbKA3HQdSOWkbvakE5pi2F2nUNkZrJOWlJRj/fLv6m+iQWVjS/1eGDfKFTJEr3m5L8u5jgvy5qhnExc9baI/eTO/a8FBCUy4EKMv6OL+1fZmXwIdSV1jtA+1ItlySVeZG+7mMc8rlb96HwTrpQ15v9256pkDMsFlvmDwGW/10yN/95KUz9H/9q7jRkEFrcJGl11p/+A5r2w0ZV3pGbPB6NnjuSE/CTHq3bdlDttmSTQ0jcfz7+wqQzppZlrLCYu+fIKUiCOBmHnMomC/RkUBpD+dJNrqR7c560CAwEAAaOCAQUwggEBMB8GA1UdIwQYMBaAFBg3LrFmVu3491fKzSAkNuCZFletMAkGA1UdEwQCMAAwEQYJYIZIAYb4QgEBBAQDAgWgMCQGCWCGSAGG+EIBDQQXFhVTU1NEIHRlc3QgQ2VydGlmaWNhdGUwHQYDVR0OBBYEFLqZmpTbOgVz8KSsrJXxwzW9+HGZMA4GA1UdDwEB/wQEAwIF4DAdBgNVHSUEFjAUBggrBgEFBQcDAgYIKwYBBQUHAwQwTAYDVR0RBEUwQ4Ehc3NzZC1kZXZlbEBsaXN0cy5mZWRvcmFob3N0ZWQub3Jnhh5odHRwczovL2dpdGh1Yi5jb20vU1NTRC9zc3NkLy8wDQYJKoZIhvcNAQELBQADggIBALNKTUZSwDLS13AVd2oJwkcnAPItwFlJwNBZurnOlkqpddFrGrY1s3AR+1V8ioZc/gy50ccf6oAsfg0M4uv3437BIwHAYS1nif+3fLWtmF6rnleM38G2xU1CTu4tlt5CQ74Lyg9wkL9qeSCxnzQyKdC3O3AkHnU9Go0uVEx6lAsFunFaRNvgcfZb98DVxCONkTvBDBiu4BngySw450kVz8SVboBNgxGyBH89lFH/D37+ATMLXo3lXZJZ0AW3b5jpffpeBlls82VAuK+7Wh43p/+1s0Mh/zEiX8UjRO1FVU0U3IRb/V6HnOTYpty8KUKciPBKUvrfyaHAyYLdbyhmWRl3wJn8NxkeLid7XLulapGxe8GBJ3WvWdaN/xtHICer68Ib8QnqbHXQE2CPsFzUZ846ZpYIlPpOQ7LJQmImAJzTaoGTjlVDWOcg3sVEjDBSw2H49Dac8RBEQI5XbSXQPutfEpB3U7uCKLfBZGf6+wZPws3XiMTtxx9v9apypVJtINBl3SILTWLZnwjhibMRr2MHMGMJOm6zju/pO13VSmZTuHpw/WIbIQrFOybnB8LV4DNlR8NrGaZ8hrjTlY+2wS2NpY2rlz+J6FmU9kdH9CpMqYUBWDrbj4K7nPA9Wgt9/4AL9rKKONtM1ugKxtB/tsO6lX/DFkRaso0fVX3S+Pey
-----------------------------------
Added certificates to user "scuser"
-----------------------------------
  User login: scuser
  Certificate: MIIE/TCCAuWgAwIBAgIBATANBgkqhkiG9w0BAQsFADA6MQ0wCwYDVQQKDARTU1NEMRIwEAYDVQQLDAlTU1NEIHRlc3QxFTATBgNVBAMMDFNTU0QgdGVzdCBDQTAeFw0yMzA3MTkxMTE5MjVaFw0yNDAyMDQxMTE5MjVaMEExDTALBgNVBAoMBFNTU0QxEjAQBgNVBAsMCVNTU0QgdGVzdDEcMBoGA1UEAwwTU1NTRCB0ZXN0IGNlcnQgMDAwMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANfzGCUs/4Pnh0ZX+Wd2ZDthGDUYO5pmKXOH+N8C9remxJgOefV/3CbKA3HQdSOWkbvakE5pi2F2nUNkZrJOWlJRj/fLv6m+iQWVjS/1eGDfKFTJEr3m5L8u5jgvy5qhnExc9baI/eTO/a8FBCUy4EKMv6OL+1fZmXwIdSV1jtA+1ItlySVeZG+7mMc8rlb96HwTrpQ15v9256pkDMsFlvmDwGW/10yN/95KUz9H/9q7jRkEFrcJGl11p/+A5r2w0ZV3pGbPB6NnjuSE/CTHq3bdlDttmSTQ0jcfz7+wqQzppZlrLCYu+fIKUiCOBmHnMomC/RkUBpD+dJNrqR7c560CAwEAAaOCAQUwggEBMB8GA1UdIwQYMBaAFBg3LrFmVu3491fKzSAkNuCZFletMAkGA1UdEwQCMAAwEQYJYIZIAYb4QgEBBAQDAgWgMCQGCWCGSAGG+EIBDQQXFhVTU1NEIHRlc3QgQ2VydGlmaWNhdGUwHQYDVR0OBBYEFLqZmpTbOgVz8KSsrJXxwzW9+HGZMA4GA1UdDwEB/wQEAwIF4DAdBgNVHSUEFjAUBggrBgEFBQcDAgYIKwYBBQUHAwQwTAYDVR0RBEUwQ4Ehc3NzZC1kZXZlbEBsaXN0cy5mZWRvcmFob3N0ZWQub3Jnhh5odHRwczovL2dpdGh1Yi5jb20vU1NTRC9zc3NkLy8wDQYJKoZIhvcNAQELBQADggIBALNKTUZSwDLS13AVd2oJwkcnAPItwFlJwNBZurnOlkqpddFrGrY1s3AR+1V8ioZc/gy50ccf6oAsfg0M4uv3437BIwHAYS1nif+3fLWtmF6rnleM38G2xU1CTu4tlt5CQ74Lyg9wkL9qeSCxnzQyKdC3O3AkHnU9Go0uVEx6lAsFunFaRNvgcfZb98DVxCONkTvBDBiu4BngySw450kVz8SVboBNgxGyBH89lFH/D37+ATMLXo3lXZJZ0AW3b5jpffpeBlls82VAuK+7Wh43p/+1s0Mh/zEiX8UjRO1FVU0U3IRb/V6HnOTYpty8KUKciPBKUvrfyaHAyYLdbyhmWRl3wJn8NxkeLid7XLulapGxe8GBJ3WvWdaN/xtHICer68Ib8QnqbHXQE2CPsFzUZ846ZpYIlPpOQ7LJQmImAJzTaoGTjlVDWOcg3sVEjDBSw2H49Dac8RBEQI5XbSXQPutfEpB3U7uCKLfBZGf6+wZPws3XiMTtxx9v9apypVJtINBl3SILTWLZnwjhibMRr2MHMGMJOm6zju/pO13VSmZTuHpw/WIbIQrFOybnB8LV4DNlR8NrGaZ8hrjTlY+2wS2NpY2rlz+J6FmU9kdH9CpMqYUBWDrbj4K7nPA9Wgt9/4AL9rKKONtM1ugKxtB/tsO6lX/DFkRaso0fVX3S+Pey
```

Now back to the client. Tell SSSD to use the SoftHSM2 config:
```
echo 'SOFTHSM2_CONF=/root/sssd/x86_64/src/tests/test_CA/softhsm2_one.conf' >> /etc/sysconfig/sssd
```
and enable certficate authentication and disable the verification on the client for a start by adding
```
pam_cert_auth = True
pam_cert_db_path = /var/lib/ipa-client/pki/ca-bundle.pem
```
to the [pam] section and
```
certificate_verification = no_verification
```
to the [sssd] section of sssd.conf (yes, `pam_cert_db_path` is needed for `no_verification` as well).

After restarting SSSD you should already see the Smartcard PIN prompt:
```
# su - scuser@ipa.test
Last login: Wed Jul 19 11:44:41 UTC 2023 on pts/0
$ su - scuser@ipa.test
PIN for SSSD Test Token: 
```
The PIN of the test tokens is `123456` but currently authentication will fail since verification is only disabled on the client (i.e. local Smartcard authentication would already work) but cannot be disabled for PKINIT which is used here by default.

To distribute the test CA certificate copy `/root/sssd/x86_64/src/tests/test_CA/SSSD_test_CA.pem` to the IPA server and call on the server:
```
]# ipa-cacert-manage install ./SSSS_test_CA.pem -t CT,C,C
Installing CA certificate, please wait
Verified CN=SSSD test CA,OU=SSSD test,O=SSSD
CA certificate successfully installed
The ipa-cacert-manage command was successful
# ipa-certupdate
Systemwide CA database updated.
Systemwide CA database updated.
The ipa-certupdate command was successful
# systemctl restart krb5kdc.service
```
On the client the IPA CA certificate database should be updated as well, since it looks like `ipa-certupdate` has `/etc/krb5.keytab` hardcoded, the soft-link as first step is needed:
```
# ln -s /enrollment/ipa.keytab /etc/krb5.keytab
# ipa-certupdate 
Systemwide CA database updated.
Systemwide CA database updated.
The ipa-certupdate command was successful
```
Now, after restarting SSSD, the Smartcard login with PKINIT should be successful:
```
# su - scuser@ipa.test
Last login: Wed Jul 19 11:47:15 UTC 2023 on pts/0
Last failed login: Wed Jul 19 11:48:35 UTC 2023 on pts/0
There was 1 failed login attempt since the last successful login.
$ su - scuser@ipa.test
PIN for SSSD Test Token: 
Last login: Wed Jul 19 12:05:28 UTC 2023 on pts/0
-sh-5.2$ klist
Ticket cache: KCM:858600005:21621
Default principal: scuser@IPA.TEST

Valid starting     Expires            Service principal
07/19/23 12:05:33  07/20/23 12:05:23  krbtgt/IPA.TEST@IPA.TEST
```


HTH

bye,
Sumit

justin-stephenson commented:
> I extended my test-matrix a bit and found that the unconditional removal of the cert info data is causing an issue if the server side can handled Smartcard authentication. In this case, if `enable:passkey` is set I would expect that all server side methods (including Smartcart) will be available and local passkey will be added is possible. For this the removal need another condition and should only happen during the evaluation of the `enable:...` option, see [sumit-bose@d330e0e](https://github.com/sumit-bose/sssd/commit/d330e0e64552d2130fb72e739c566d6207ac523d) .
> 
> bye, Sumit

Incorporated https://github.com/sumit-bose/sssd/commit/d330e0e64552d2130fb72e739c566d6207ac523d  into this PR.

sumit-bose reviewed (approved):
Hi,

thank you, I didn't came across other issues or unexpected behavior, so ACK,

Thanks for your hard work and your patience.

bye,
Sumit

ikerexxe commented:
> Covscan shows the following warning, but I don't see it related to my patches. Please correct me if i'm mistaken.
> 
> ```
>  Warning: Error: CLANG_WARNING:
> sssd-pr6696/src/sss_client/pam_sss.c:2966:21: warning[unix.Malloc]: Potential leak of memory pointed to by 'pi.pam_authtok'
> # 2964|                   ret = get_authtok_for_authentication(pamh, &pi, flags);
> # 2965|                   if (ret != PAM_SUCCESS) {
> # 2966|->                     D(("failed to get authentication token: %s",
> # 2967|                          pam_strerror(pamh, ret)));
> # 2968|                       return ret;
> ```

I agree with you.

I tried to follow Sumit's steps to use Smartcard authentication, but I must be doing something wrong because I never get the Smartcard PIN prompt. In any case, everything else is working correctly, and Sumit already tested that path, so I'm approving this PR.

Thank you for taking care of it Justin! Excellent work

alexey-tikhonov commented:
@ikerexxe, do you plan to pull this into F39 via patches?

ikerexxe commented:
> @ikerexxe, do you plan to pull this into F39 via patches?

That's an option, but if you think we should release a new sssd version then I'm fine with it. By the way, if we are to release a new version then there's another patch missing from it.

ikerexxe commented:
> So far it wasn't planned. My question was more like: are those patches needed in 2.9 branch?

Yes, I think it would be better. At least if something happens and we need to release a new version the patches will already be there.

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/6696

* `master`
    * 43d89dd2d9d9c86ecd487067a6bbdf1fbf1513bb - PAM: Fail empty password in passkey fallback
    * d019132bd44e25b841e0917c034140be67de9a77 - Add local auth policy
    * f3f7a4ce11a91f723d4f729858ebb946fdd6c5e2 - Change "non_kerberos" to "local" authentication
    * bfab4907535742128d7140ba1ad858565f70fe3a - man: clarify passkey PIN prompt
* `sssd-2-9`
    * 16f12efd9806dd903c6d03d9f48dba9a34a7e9d5 - PAM: Fail empty password in passkey fallback
    * 5b575fcb62c1e0ea98905fb5e942d437f7ef8646 - Add local auth policy
    * b8b75abebca6234f210b6d19ccabfb4eb36e217c - Change "non_kerberos" to "local" authentication
    * 8c1b5c472eb700dfd579e16145d8bc1cdcaa0ea0 - man: clarify passkey PIN prompt
